man/gsm_explode.3
.\"
.\" Copyright 1992 by Jutta Degener and Carsten Bormann, Technische
.\" Universitaet Berlin.  See the accompanying file "LICENSE" for
.\" details.  THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.
.\"
.PU
.TH GSM_EXPLODE 3 
.SH NAME
gsm_explode, gsm_implode \(em GSM\ 06.10 supplementary
functions for testing
.SH SYNOPSIS
#include "gsm.h"
.PP
void gsm_explode(g, frame, xframe)
.br
gsm        g;
.br
gsm_frame  frame;
.br
gsm_signal xframe[ 76 ];
.PP
void gsm_implode(g, xframe, frame)
.br
gsm        g;
.br
gsm_signal xframe[ 76 ];
.br
gsm_frame  frame;
.SH "DESCRIPTION"
Gsm is an implementation of the final draft GSM 06.10
standard for full-rate speech transcoding.
Test data for implementations of this particular document
can be bought and used to verify an implementation.
.PP
The encoded test data uses a format different from what
one would use to transmit frames with the least number
of bits.
Gsm_explode() and gsm_implode() convert between the
internal, small, 33-byte format and the 76-word format
used by the test data.
.PP
.SH "RETURN VALUE"
gsm_explode() returns -1 if the passed frame is invalid, else 0.
.SH BUGS
Please direct bug reports to jutta@pobox.com and cabo@tzi.org.
.SH "SEE ALSO"
gsm(3)


ChangeLog
Thu Aug  4 18:47:10 2022	Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 22
	Remove unused, and occasionally out of date, VERSION variable
	from Makefile.

Sun Jul 31 16:17:59 2022	Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 21
	Use $(LN), not ln, in the Makefile rules for the installed binary
 	of untoast and tcat.
	(Thanks to Carlo Cabrera for pointing that out!)

Mon Jul  4 10:34:18 2022

	* Release 1.0 Patchlevel 20
	Use $(RMFLAGS) with the rms in the Makefile; add -f to $(RMFLAGS) to
        avoid spurious error messages during build.

Fri Mar 27 15:18:06  2020

	* Release 1.0 Patchlevel 19
	Make it easier to include gsm.h from C++ by wrapping it with macros. 
	Throw a #define _POSIX_C_SOURCE 200809L into toast.h to make it
	possible for Linux glibc to just work.
	(Thanks to enh@google.com for the suggestion; errors are mine.)
	Switch from compress/.Z to gzip/.gz 

Sat May  5 21:48:20  2018

	* Release 1.0 Patchlevel 18
	Remove the MacOS "quarantine" files that I didn't know I'd shipped.
	Remove some trailing spaces.
	Bump up the patchlevel in inc/gsm.h (I'd forgotten that in release 17.)

Tue Jun 27 17:59:30  2017

	* Release 1.0 Patchlevel 17
	New email addresses in the README.
	Avoid left-shifting signed integers.
	Don't print longwords using %d or %ld, cast them to (long) first.
	Don't compare longwords as longs in addtest, we only care for the
		lower 32 bits. (The math works, but the test failed.)

Sat Sep 17 14:50:55  2016

	* Release 1.0 Patchlevel 16
	New email addresses in the manual pages (patch 16)
	Removed a stray .orig file (in patch 15) 

Sun Apr  5  04:53:07 2009       Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 13
	COPYRIGHT: exploring the deeper meaning and cultural
	history of the word "any".

Fri Jun  2 06:27:56 2006	Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 12
	src/code.c: removed modifiable static storage that made the library
	 	non-reentrant.  (Thanks to Paul Gibbs for pointing out the bug.)

Sat Mar 18 11:48:28 2006	Jutta Degener (jutta@pobox.com)

	* Release 1.0 Patchlevel 11
	src/gsm_implode.c: fix WAV49 implode
	move source archive location

Fri Jul  5 19:26:37 1996 	Jutta Degener (jutta@cs.tu-berlin.de)

	* Release 1.0 Patchlevel 10
	src/toast_alaw.c: exchanged A-law tables for something
		slightly more A-law.

Tue Jul  2 12:18:20 1996  Jutta Degener (jutta@cs.tu-berlin.de)

	* Release 1.0 Patchlevel 9
	src/long_term.c: in FLOAT_MUL mode, an array was accessed past its end
	src/gsm_option.c: three options related to WAV #49 packing
	src/gsm_encode.c: support WAV #49-style encoding.
	src/gsm_decode.c: support WAV #49-style decoding.
	tls/sour.c: generate the WAV bit shifting code, encode
	tls/ginger.c: generate the WAV bit shifting code, decode
	The WAV code goes back to an inofficial patch #8 that
	Jeff Chilton sent us (hence the jump from 7 to 9).
	src/toast.c: add _fsetmode() calls to set stdin/stdout to
		binary (from an OS/2 port by Arnd Gronenberg.)

Tue Mar  7 01:55:10 1995  Jutta Degener (jutta@cs.tu-berlin.de)

	* Release 1.0 Patchlevel 7
	src/long_term.c: Yet another 16-bit overflow
	src/toast.c: -C option to toast, cuts LPC time
	src/gsm_option.c: corresponding LPC_CUT option to GSM library

Fri Dec 30 23:33:50 1994  Jutta Degener (jutta@cs.tu-berlin.de)

        * Release 1.0 Patchlevel 6
        src/lpc.c: fixed 16-bit addition overflow in Autocorrelation code
        src/add.c: gsm_L_asl should fall back on gsm_L_asr, not gsm_asr

Mon Nov 28 20:49:57 1994  Jutta Degener (jutta@cs.tu-berlin.de)
	
	* Release 1.0 Patchlevel 5
	src/toast_audio.c: initialization should return -1 on error
	src/gsm_destroy.c: #include configuration header file
	src/add.c: gsm_sub should cast its parameters to longword
	man/*: bug reports to {jutta,cabo}@cs.tu-berlin.de, not to toast@tub
	inc/private.h: longword long by default, not int
	inc/toast.h: read/write fopen modes "rb" and "wb", not just "r"
	src/toast.c: better (or different, anyway) error handling in process()

Tue May 10 19:41:34 1994  Jutta Degener (jutta at kugelbus)
	
	* Release 1.0 Patchlevel 4
	inc/private.h: GSM_ADD should cast to ulongword, not to unsigned.
	src/long_term.c: missing cast to longword.
	add-test/add_test.c: Test macros too, not only functions,
	thanks to Simao Ferraz de Campos Neto, simao@dragon.cpqd.ansp.br
	General cleanup: remove unused variables, add function prototypes.

Tue Jan 25 22:53:40 1994  Jutta Degener (jutta at kugelbus)

	* Release 1.0 Patchlevel 3
	changed rpe.c's STEP macro to work with 16-bit integers,
	thanks to Dr Alex Lee (alexlee@solomon.technet.sg);
	removed non-fatal bugs from add-test.dta, private.h
	and toast_audio.c, thanks to P. Emanuelsson.

Fri Jan 29 19:02:12 1993  Jutta Degener  (jutta at kraftbus)
	
	* Release 1.0 Patchlevel 2
	fixed L_add(0,-1) in src/add.c and inc/private.h,
	thanks to Raphael Trommer at AT&T Bell Laboratories;
	various other ANSI C compatibility details

Fri Oct 30 17:58:54 1992  Jutta Degener  (jutta at kraftbus)

	* Release 1.0 Patchlevel 1
	Switched uid/gid in toast's [f]chown calls.

Wed Oct 28 14:12:35 1992  Carsten Bormann  (cabo at kubus)

	* Release 1.0: released
	Copyright 1992 by Jutta Degener and Carsten Bormann, Technische
	Universitaet Berlin.  See the accompanying file "LICENSE" for
	details.  THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.



src/gsm_explode.c
/*
 * Copyright 1992 by Jutta Degener and Carsten Bormann, Technische
 * Universitaet Berlin.  See the accompanying file "LICENSE" for
 * details.  THERE IS ABSOLUTELY NO WARRANTY FOR THIS SOFTWARE.
 */

/* $Header: /tmp_amd/presto/export/kbs/jutta/src/gsm/RCS/gsm_explode.c,v 1.2 1996/07/02 14:32:42 jutta Exp jutta $ */

#include "private.h"
#include "gsm.h"
#include "proto.h"

int gsm_explode P3((s, c, target), gsm s, gsm_byte * c, gsm_signal * target)
{
#	define	LARc	target
#	define	Nc	*((gsm_signal (*) [17])(target + 8))
#	define	bc	*((gsm_signal (*) [17])(target + 9))
#	define	Mc	*((gsm_signal (*) [17])(target + 10))
#	define	xmaxc	*((gsm_signal (*) [17])(target + 11))


#ifdef WAV49
	if (s->wav_fmt) {

		uword sr = 0;

		if (s->frame_index == 1) {

			sr = *c++;
			LARc[0] = sr & 0x3f;  sr >>= 6;
			sr |= (uword)*c++ << 2;
			LARc[1] = sr & 0x3f;  sr >>= 6;
			sr |= (uword)*c++ << 4;
			LARc[2] = sr & 0x1f;  sr >>= 5;
			LARc[3] = sr & 0x1f;  sr >>= 5;
			sr |= (uword)*c++ << 2;
			LARc[4] = sr & 0xf;  sr >>= 4;
			LARc[5] = sr & 0xf;  sr >>= 4;
			sr |= (uword)*c++ << 2;			/* 5 */
			LARc[6] = sr & 0x7;  sr >>= 3;
			LARc[7] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 4;
			Nc[0] = sr & 0x7f;  sr >>= 7;
			bc[0] = sr & 0x3;  sr >>= 2;
			Mc[0] = sr & 0x3;  sr >>= 2;
			sr |= (uword)*c++ << 1;
			xmaxc[0] = sr & 0x3f;  sr >>= 6;
#undef	xmc
#define	xmc	(target + 12)
			xmc[0] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			xmc[1] = sr & 0x7;  sr >>= 3;
			xmc[2] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[3] = sr & 0x7;  sr >>= 3;
			xmc[4] = sr & 0x7;  sr >>= 3;
			xmc[5] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;			/* 10 */
			xmc[6] = sr & 0x7;  sr >>= 3;
			xmc[7] = sr & 0x7;  sr >>= 3;
			xmc[8] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			xmc[9] = sr & 0x7;  sr >>= 3;
			xmc[10] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[11] = sr & 0x7;  sr >>= 3;
			xmc[12] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 4;
			Nc[1] = sr & 0x7f;  sr >>= 7;
			bc[1] = sr & 0x3;  sr >>= 2;
			Mc[1] = sr & 0x3;  sr >>= 2;
			sr |= (uword)*c++ << 1;
			xmaxc[1] = sr & 0x3f;  sr >>= 6;
#undef	xmc
#define	xmc	(target + 29 - 13)

			xmc[13] = sr & 0x7;  sr >>= 3;
			sr = *c++;				/* 15 */
			xmc[14] = sr & 0x7;  sr >>= 3;
			xmc[15] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[16] = sr & 0x7;  sr >>= 3;
			xmc[17] = sr & 0x7;  sr >>= 3;
			xmc[18] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;
			xmc[19] = sr & 0x7;  sr >>= 3;
			xmc[20] = sr & 0x7;  sr >>= 3;
			xmc[21] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			xmc[22] = sr & 0x7;  sr >>= 3;
			xmc[23] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[24] = sr & 0x7;  sr >>= 3;
			xmc[25] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 4;			/* 20 */
			Nc[2] = sr & 0x7f;  sr >>= 7;
			bc[2] = sr & 0x3;  sr >>= 2;
			Mc[2] = sr & 0x3;  sr >>= 2;
			sr |= (uword)*c++ << 1;
			xmaxc[2] = sr & 0x3f;  sr >>= 6;

#undef	xmc
#define	xmc	(target + 46 - 26)

			xmc[26] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			xmc[27] = sr & 0x7;  sr >>= 3;
			xmc[28] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[29] = sr & 0x7;  sr >>= 3;
			xmc[30] = sr & 0x7;  sr >>= 3;
			xmc[31] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;
			xmc[32] = sr & 0x7;  sr >>= 3;
			xmc[33] = sr & 0x7;  sr >>= 3;
			xmc[34] = sr & 0x7;  sr >>= 3;
			sr = *c++;				/* 25 */
			xmc[35] = sr & 0x7;  sr >>= 3;
			xmc[36] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[37] = sr & 0x7;  sr >>= 3;
			xmc[38] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 4;
			Nc[3] = sr & 0x7f;  sr >>= 7;
			bc[3] = sr & 0x3;  sr >>= 2;
			Mc[3] = sr & 0x3;  sr >>= 2;
			sr |= (uword)*c++ << 1;
			xmaxc[3] = sr & 0x3f;  sr >>= 6;
#undef	xmc
#define	xmc	(target + 63 - 39)

			xmc[39] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			xmc[40] = sr & 0x7;  sr >>= 3;
			xmc[41] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;			/* 30 */
			xmc[42] = sr & 0x7;  sr >>= 3;
			xmc[43] = sr & 0x7;  sr >>= 3;
			xmc[44] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;
			xmc[45] = sr & 0x7;  sr >>= 3;
			xmc[46] = sr & 0x7;  sr >>= 3;
			xmc[47] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			xmc[48] = sr & 0x7;  sr >>= 3;
			xmc[49] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[50] = sr & 0x7;  sr >>= 3;
			xmc[51] = sr & 0x7;  sr >>= 3;

			s->frame_chain = sr & 0xf;
		}
		else {
			sr = s->frame_chain;
			sr |= (uword)*c++ << 4;			/* 1 */
			LARc[0] = sr & 0x3f;  sr >>= 6;
			LARc[1] = sr & 0x3f;  sr >>= 6;
			sr = *c++;
			LARc[2] = sr & 0x1f;  sr >>= 5;
			sr |= (uword)*c++ << 3;
			LARc[3] = sr & 0x1f;  sr >>= 5;
			LARc[4] = sr & 0xf;  sr >>= 4;
			sr |= (uword)*c++ << 2;
			LARc[5] = sr & 0xf;  sr >>= 4;
			LARc[6] = sr & 0x7;  sr >>= 3;
			LARc[7] = sr & 0x7;  sr >>= 3;
			sr = *c++;				/* 5 */
			Nc[0] = sr & 0x7f;  sr >>= 7;
			sr |= (uword)*c++ << 1;
			bc[0] = sr & 0x3;  sr >>= 2;
			Mc[0] = sr & 0x3;  sr >>= 2;
			sr |= (uword)*c++ << 5;
			xmaxc[0] = sr & 0x3f;  sr >>= 6;
#undef	xmc
#define	xmc	(target + 12)
			xmc[0] = sr & 0x7;  sr >>= 3;
			xmc[1] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;
			xmc[2] = sr & 0x7;  sr >>= 3;
			xmc[3] = sr & 0x7;  sr >>= 3;
			xmc[4] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			xmc[5] = sr & 0x7;  sr >>= 3;
			xmc[6] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;			/* 10 */
			xmc[7] = sr & 0x7;  sr >>= 3;
			xmc[8] = sr & 0x7;  sr >>= 3;
			xmc[9] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;
			xmc[10] = sr & 0x7;  sr >>= 3;
			xmc[11] = sr & 0x7;  sr >>= 3;
			xmc[12] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			Nc[1] = sr & 0x7f;  sr >>= 7;
			sr |= (uword)*c++ << 1;
			bc[1] = sr & 0x3;  sr >>= 2;
			Mc[1] = sr & 0x3;  sr >>= 2;
			sr |= (uword)*c++ << 5;
			xmaxc[1] = sr & 0x3f;  sr >>= 6;
#undef	xmc
#define	xmc	(target + 29 - 13)

			xmc[13] = sr & 0x7;  sr >>= 3;
			xmc[14] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;			/* 15 */
			xmc[15] = sr & 0x7;  sr >>= 3;
			xmc[16] = sr & 0x7;  sr >>= 3;
			xmc[17] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			xmc[18] = sr & 0x7;  sr >>= 3;
			xmc[19] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[20] = sr & 0x7;  sr >>= 3;
			xmc[21] = sr & 0x7;  sr >>= 3;
			xmc[22] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;
			xmc[23] = sr & 0x7;  sr >>= 3;
			xmc[24] = sr & 0x7;  sr >>= 3;
			xmc[25] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			Nc[2] = sr & 0x7f;  sr >>= 7;
			sr |= (uword)*c++ << 1;			/* 20 */
			bc[2] = sr & 0x3;  sr >>= 2;
			Mc[2] = sr & 0x3;  sr >>= 2;
			sr |= (uword)*c++ << 5;
			xmaxc[2] = sr & 0x3f;  sr >>= 6;
#undef	xmc
#define	xmc	(target + 46 - 26)
			xmc[26] = sr & 0x7;  sr >>= 3;
			xmc[27] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;	
			xmc[28] = sr & 0x7;  sr >>= 3;
			xmc[29] = sr & 0x7;  sr >>= 3;
			xmc[30] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			xmc[31] = sr & 0x7;  sr >>= 3;
			xmc[32] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[33] = sr & 0x7;  sr >>= 3;
			xmc[34] = sr & 0x7;  sr >>= 3;
			xmc[35] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;			/* 25 */
			xmc[36] = sr & 0x7;  sr >>= 3;
			xmc[37] = sr & 0x7;  sr >>= 3;
			xmc[38] = sr & 0x7;  sr >>= 3;
			sr = *c++;
			Nc[3] = sr & 0x7f;  sr >>= 7;
			sr |= (uword)*c++ << 1;		
			bc[3] = sr & 0x3;  sr >>= 2;
			Mc[3] = sr & 0x3;  sr >>= 2;
			sr |= (uword)*c++ << 5;
			xmaxc[3] = sr & 0x3f;  sr >>= 6;

#undef	xmc
#define	xmc	(target + 63 - 39)

			xmc[39] = sr & 0x7;  sr >>= 3;
			xmc[40] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;
			xmc[41] = sr & 0x7;  sr >>= 3;
			xmc[42] = sr & 0x7;  sr >>= 3;
			xmc[43] = sr & 0x7;  sr >>= 3;
			sr = *c++;				/* 30 */
			xmc[44] = sr & 0x7;  sr >>= 3;
			xmc[45] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 2;
			xmc[46] = sr & 0x7;  sr >>= 3;
			xmc[47] = sr & 0x7;  sr >>= 3;
			xmc[48] = sr & 0x7;  sr >>= 3;
			sr |= (uword)*c++ << 1;
			xmc[49] = sr & 0x7;  sr >>= 3;
			xmc[50] = sr & 0x7;  sr >>= 3;
			xmc[51] = sr & 0x7;  sr >>= 3;
		}
	}
	else
#endif
	{
	/* GSM_MAGIC  = (*c >> 4) & 0xF; */

	if (((*c >> 4) & 0x0F) != GSM_MAGIC) return -1;

	LARc[0]  = (*c++ & 0xF) << 2;		/* 1 */
	LARc[0] |= (*c >> 6) & 0x3;
	LARc[1]  = *c++ & 0x3F;
	LARc[2]  = (*c >> 3) & 0x1F;
	LARc[3]  = (*c++ & 0x7) << 2;
	LARc[3] |= (*c >> 6) & 0x3;
	LARc[4]  = (*c >> 2) & 0xF;
	LARc[5]  = (*c++ & 0x3) << 2;
	LARc[5] |= (*c >> 6) & 0x3;
	LARc[6]  = (*c >> 3) & 0x7;
	LARc[7]  = *c++ & 0x7;

	Nc[0]  = (*c >> 1) & 0x7F;

	bc[0]  = (*c++ & 0x1) << 1;
	bc[0] |= (*c >> 7) & 0x1;

	Mc[0]  = (*c >> 5) & 0x3;

	xmaxc[0]  = (*c++ & 0x1F) << 1;
	xmaxc[0] |= (*c >> 7) & 0x1;

#undef	xmc
#define	xmc	(target + 12)

	xmc[0]  = (*c >> 4) & 0x7;
	xmc[1]  = (*c >> 1) & 0x7;
	xmc[2]  = (*c++ & 0x1) << 2;
	xmc[2] |= (*c >> 6) & 0x3;
	xmc[3]  = (*c >> 3) & 0x7;
	xmc[4]  = *c++ & 0x7;
	xmc[5]  = (*c >> 5) & 0x7;
	xmc[6]  = (*c >> 2) & 0x7;
	xmc[7]  = (*c++ & 0x3) << 1;		/* 10 */
	xmc[7] |= (*c >> 7) & 0x1;
	xmc[8]  = (*c >> 4) & 0x7;
	xmc[9]  = (*c >> 1) & 0x7;
	xmc[10]  = (*c++ & 0x1) << 2;
	xmc[10] |= (*c >> 6) & 0x3;
	xmc[11]  = (*c >> 3) & 0x7;
	xmc[12]  = *c++ & 0x7;

	Nc[1]  = (*c >> 1) & 0x7F;

	bc[1]  = (*c++ & 0x1) << 1;
	bc[1] |= (*c >> 7) & 0x1;

	Mc[1]  = (*c >> 5) & 0x3;

	xmaxc[1]  = (*c++ & 0x1F) << 1;
	xmaxc[1] |= (*c >> 7) & 0x1;

#undef	xmc
#define	xmc	(target + 29 - 13)

	xmc[13]  = (*c >> 4) & 0x7;
	xmc[14]  = (*c >> 1) & 0x7;
	xmc[15]  = (*c++ & 0x1) << 2;
	xmc[15] |= (*c >> 6) & 0x3;
	xmc[16]  = (*c >> 3) & 0x7;
	xmc[17]  = *c++ & 0x7;
	xmc[18]  = (*c >> 5) & 0x7;
	xmc[19]  = (*c >> 2) & 0x7;
	xmc[20]  = (*c++ & 0x3) << 1;
	xmc[20] |= (*c >> 7) & 0x1;
	xmc[21]  = (*c >> 4) & 0x7;
	xmc[22]  = (*c >> 1) & 0x7;
	xmc[23]  = (*c++ & 0x1) << 2;
	xmc[23] |= (*c >> 6) & 0x3;
	xmc[24]  = (*c >> 3) & 0x7;
	xmc[25]  = *c++ & 0x7;

	Nc[2]  = (*c >> 1) & 0x7F;

	bc[2]  = (*c++ & 0x1) << 1;		/* 20 */
	bc[2] |= (*c >> 7) & 0x1;

	Mc[2]  = (*c >> 5) & 0x3;

	xmaxc[2]  = (*c++ & 0x1F) << 1;
	xmaxc[2] |= (*c >> 7) & 0x1;

#undef	xmc
#define	xmc	(target + 46 - 26)

	xmc[26]  = (*c >> 4) & 0x7;
	xmc[27]  = (*c >> 1) & 0x7;
	xmc[28]  = (*c++ & 0x1) << 2;
	xmc[28] |= (*c >> 6) & 0x3;
	xmc[29]  = (*c >> 3) & 0x7;
	xmc[30]  = *c++ & 0x7;
	xmc[31]  = (*c >> 5) & 0x7;
	xmc[32]  = (*c >> 2) & 0x7;
	xmc[33]  = (*c++ & 0x3) << 1;
	xmc[33] |= (*c >> 7) & 0x1;
	xmc[34]  = (*c >> 4) & 0x7;
	xmc[35]  = (*c >> 1) & 0x7;
	xmc[36]  = (*c++ & 0x1) << 2;
	xmc[36] |= (*c >> 6) & 0x3;
	xmc[37]  = (*c >> 3) & 0x7;
	xmc[38]  = *c++ & 0x7;

	Nc[3]  = (*c >> 1) & 0x7F;

	bc[3]  = (*c++ & 0x1) << 1;
	bc[3] |= (*c >> 7) & 0x1;

	Mc[3]  = (*c >> 5) & 0x3;

	xmaxc[3]  = (*c++ & 0x1F) << 1;
	xmaxc[3] |= (*c >> 7) & 0x1;

#undef	xmc
#define	xmc	(target + 63 - 39)

	xmc[39]  = (*c >> 4) & 0x7;
	xmc[40]  = (*c >> 1) & 0x7;
	xmc[41]  = (*c++ & 0x1) << 2;
	xmc[41] |= (*c >> 6) & 0x3;
	xmc[42]  = (*c >> 3) & 0x7;
	xmc[43]  = *c++ & 0x7;			/* 30  */
	xmc[44]  = (*c >> 5) & 0x7;
	xmc[45]  = (*c >> 2) & 0x7;
	xmc[46]  = (*c++ & 0x3) << 1;
	xmc[46] |= (*c >> 7) & 0x1;
	xmc[47]  = (*c >> 4) & 0x7;
	xmc[48]  = (*c >> 1) & 0x7;
	xmc[49]  = (*c++ & 0x1) << 2;
	xmc[49] |= (*c >> 6) & 0x3;
	xmc[50]  = (*c >> 3) & 0x7;
	xmc[51]  = *c & 0x7;			/* 33 */
	}

	return 0;
}
